### PR TITLE
pythonPackages.bcrypt: use 3.1.x for python older than 3.6

### PR DESCRIPTION
**Pull request by mvnetbiz**
###### Motivation for this change
pythonPackages.bcrypt was disabled for python older than 3.6: https://github.com/NixOS/nixpkgs/commit/87371b8f6338af8398410443868c1f7750840c4b

We still have some applications that cannot be upgraded from python 2 that require python-bcrypt
mysql-workbench, syncthing-gtk: https://github.com/NixOS/nixpkgs/issues/97642

ZHF: #97479

###### Things done
Added back pythonPackages.bcrypt 3.1.7 only for python older than 3.6

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by drewrisinger** [comment]
@mvnetbiz the standard way of doing this is overriding the package set to create a custom python installation with the version that you want. See for example https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/tools/prospector/default.nix.

That is, for ``mysql_workbench.nix``, you add something like 
```
let
  py2 = python2.override {
    packageOverrides = self: super:
      bcrypt = super.bcrypt.overridePythonAttrs(oldAttrs: rec {
        version = "3.1.X";
        src = oldAttrs.src.override {
          inherit version;
          sha256 = "...";
        };
        disabled = false;
      };
    });
  };
in
  ...

**Comment by mvnetbiz** [comment]
Since it is deprecated it seems like it would make sense to keep the definition in the package that requires it, so it can be removed once it isn't needed any more, but this is used by at least 2 used packages, also the expression might deviate significantly from the new version, would that end up being too verbose keeping in a per-use override? Would it be better to add a separate pythonPackages.bcrypt_3_1, and use that in these packages?

**Comment by drewrisinger** [comment]
> Since it is deprecated it seems like it would make sense to keep the definition in the package that requires it, so it can be removed once it isn't needed any more, but this is used by at least 2 used packages, also the expression might deviate significantly from the new version, would that end up being too verbose keeping in a per-use override? Would it be better to add a separate pythonPackages.bcrypt_3_1, and use that in these packages?

I took about 10 mins and looked at usages of bcrypt throughout nixpkgs. In the applications that need it, ``bcrypt`` is always overridden to a version ``~3.1.7`` (e.g. ``home-assistant``). I think this is probably the path that you should take. I wouldn't worry about long derivation overrides, most seem to be pretty short.

pinging maintainer of ``bcrypt`` @domenkozar

**Comment by mvnetbiz** [comment]
IMHO, in that example, it is bcrypt's reverse dependency (home-assistant) breaking compatibility with it, so it makes sense to override it there. The reason for this patch is that bcrypt itself, not it's user, is dropping compatibility with python < 3.6, so anything requiring bcrypt on python2 will have to use this version, so I thought it made sense to have a more general patch. I can make the PRs for mysql-workbench, syncthing-gtk, (poretools?) if this gets closed.

**Comment by mvnetbiz** [comment]
There are many modules defined this way in pkgs/top-level/python-packages.nix, at least the first 4 I looked at result from the same upstream dropping of old python version, this doesn't seem too atypical.

**Comment by jonringer** [comment]
@drewrisinger that's only if it's a python application, if he still needs python2 to work, this is the correct method. We do it for packages such as numpy.

**Comment by jonringer** [comment]
also, please disable python2Packages.trytond:
```
  Processing ./trytond-5.6.5-py2-none-any.whl
  ERROR: Package 'trytond' requires a different Python: 2.7.18 not in '>=3.5'
```

**Comment by mvnetbiz** [comment]
Should that be in the same PR? It's not caused by this branch, it fails to build on master too, just with:
```nix-build . -A pkgs.python2Packages.trytond
error: bcrypt-3.2.0 not supported for interpreter python2.7
(use '--show-trace' to show detailed location information)
```
instead of trying to build and fail.

File: pkgs/development/python-modules/bcrypt/3_1.nix [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+buildPythonPackage rec {
+  version = "3.1.7";
+  pname = "bcrypt";
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "0b0069c752ec14172c5f78208f1863d7ad6755a6fae6fe76ec2c80d13be41e42";
+  };
+  buildInputs = [ pycparser mock pytest py ];
```
Comment (drewrisinger) [inline pkgs/development/python-modules/bcrypt/3_1.nix:14]:
```suggestion
  checkInputs = [ pytestCheckHook ];
```
I don't see any reference to ``pycparser, mock, py`` in the ``requirements.txt`` or ``setup.py``, so these are probably unnecessary.

File: pkgs/development/python-modules/bcrypt/3_1.nix [hunk 2]
```diff
@@ -0,0 +1,23 @@
+{ stdenv, buildPythonPackage, isPyPy, fetchPypi
+, cffi, pycparser, mock, pytest, py, six }:
+
+with stdenv.lib;
```
Comment (drewrisinger) [inline pkgs/development/python-modules/bcrypt/3_1.nix:4]:
```suggestion
```

File: pkgs/development/python-modules/bcrypt/3_1.nix [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+  pname = "bcrypt";
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "0b0069c752ec14172c5f78208f1863d7ad6755a6fae6fe76ec2c80d13be41e42";
+  };
+  buildInputs = [ pycparser mock pytest py ];
+  propagatedBuildInputs = [ six ] ++ optional (!isPyPy) cffi;
+
+  meta = {
```
Comment (drewrisinger) [inline pkgs/development/python-modules/bcrypt/3_1.nix:17]:
```suggestion
  meta = with stdenv.lib; {
```

File: pkgs/development/python-modules/bcrypt/3_1.nix [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    maintainers = with maintainers; [ domenkozar ];
+    description = "Modern password hashing for your software and your servers";
+    license = licenses.asl20;
+    homepage = "https://github.com/pyca/bcrypt/";
```
Comment (drewrisinger) [inline pkgs/development/python-modules/bcrypt/3_1.nix:21]: These could be reordered.
```suggestion
    description = "Modern password hashing for your software and your servers";
    homepage = "https://github.com/pyca/bcrypt/";
    license = licenses.asl20;
    maintainers = with maintainers; [ domenkozar ];
```

**Review by drewrisinger — CHANGES_REQUESTED** [review]
* [x] A few unnecessary things in diff
* [x] Commits LGTM
* [x] Partial build tested: ``nix-review -p python27Packages.bcrypt -p mysql-workbench``
* [x] Maybe add ``disabled`` to ``python-modules/bcrypt/default.nix`` to be explicit in the file about where it's supported??

```
https://github.com/NixOS/nixpkgs/pull/97996
2 packages built:
mysql-workbench python27Packages.bcrypt
```

**Comment by mvnetbiz** [comment]
@drewrisinger, thanks for the comments. `python-modules/bcrypt/default.nix` already has `disabled` from commit referenced in PR body

**Comment by drewrisinger** [comment]
Looks like enabling this breaks ``python27Packages.[junos-eznc, trackpy]`` b/c it's semi-adding a new package.

```
https://github.com/NixOS/nixpkgs/pull/97996
2 packages failed to build:
python27Packages.junos-eznc python27Packages.trackpy

34 packages built:
mysql-workbench poretools python27Packages.Fabric python27Packages.ansible python27Packages.awkward python27Packages.bcrypt python27Packages.bkcharts python27Packages.docker python27Packages.drms python27Packages.duckdb python27Packages.flammkuchen python27Packages.flask-bcrypt python27Packages.imbalanced-learn python27Packages.ldaptor python27Packages.mapsplotlib python27Packages.moto python27Packages.ncclient python27Packages.pandas python27Packages.paramiko python27Packages.passlib python27Packages.privacyidea-ldap-proxy python27Packages.pwntools python27Packages.pysftp python27Packages.pytest-ansible python27Packages.robotframework-sshlibrary python27Packages.scp python27Packages.seaborn python27Packages.sshtunnel python27Packages.uproot python27Packages.uproot-methods python27Packages.vega_datasets python27Packages.vidstab python27Packages.yfinance syncthing-gtk
```

**Comment by mvnetbiz** [comment]
`trackpy` is marked broken, and I don't think this broke `junos-eznc`, the derivation just couldn't evaluate previously, with `error: bcrypt-3.2.0 not supported for interpreter python2.7`. It is broken for some other reason apparently.

**Review by drewrisinger — APPROVED** [review]
* [x] Diff LGTM
* [x] Commits LGTM
* [x] Builds via ``nix-review`` (mostly, see above comments). Failing packages should be addressed in follow-up ZHF PR.

**Review by jonringer — APPROVED** [review]
diff LGTM

**Pull request merged**